### PR TITLE
Feature/v2/plugin rework

### PR DESCRIPTION
**Pull request by JonGardner-RedSpace**
Tasks:
- [x] Documentation
- [x] Unit Tests
- [x] Typescript Typings
- [ ] Haxe Externs
- [ ] `npm run build:full`

### CODE REVIEW

File: src/Application.js [hunk 1]
```diff
@@ -173,6 +155,52 @@ export class Application {
     });
   }
 
+  /**
+   *
+   * @return {void}@memberof Application
```
Comment (SmithAF) [inline src/Application.js]: Formatting

File: src/Application.js [hunk 2]
```diff
@@ -173,6 +155,53 @@ export class Application {
...
+  setupPlugins() {
+    const preloads = [];
+    Application._plugins.forEach(plugin => {
+      if (!plugin.preload) {
+        return;
+      }
+
+      preloads.push(
+        plugin.preload(this).catch(function preloadFail(error) {
+          plugin.preloadFailed = true;
```
Comment (chipbell4) [inline src/Application.js]: For consistency, maybe we should go ahead and set `preloadFailed` to false on the plugin before we start the preload:

```
plugin.preloadFailed = false;

preloads.push(
...
```

File: src/Application.js [hunk 3]
```diff
@@ -173,6 +155,53 @@ export class Application {
...
+          console.warn(plugin.name, 'Preload Failed:', error);
+        })
+      );
+    });
+
+    // ~wait for all preloads to resolve
+    return Promise.all(preloads).then(() => {
+      // Remove plugins that fail to load.
+      Application._plugins = Application._plugins.filter(
+        plugin => plugin.preloadFailed != true
```
Comment (chipbell4) [inline src/Application.js]: Let's make this a stronger check: `plugin.preloadFailed !== true`.

File: src/Application.spec.js [hunk 4]
```diff
@@ -1,23 +1,57 @@
...
+  constructor() {
+    super({ name: 'failed plugin' });
+  }
 
   /** */
   preload(application) {
     this.preloadCalled = true;
     expect(application).to.be.instanceOf(Application);
-    return Promise.resolve();
+    return Promise.reject('It was rigged from the start');
```
Comment (chipbell4) [inline src/Application.spec.js]: lol

File: src/Application.js (hunk 2 above)
Comment (JonGardner-RedSpace) [reply src/Application.js]: That's actually only used in the loading process, failed plugins get removed from the list, with a warning message.

**Pull request merged**